Release 32.0.0rc1 prep

AyanSinhaMahapatra opened the pull request:
### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

AyanSinhaMahapatra committed:
Remove deprecated images from CI and release-script

The following images are deprecated in GitHub actions and Azure DevOps:

* `ubuntu-18.04` : https://github.com/actions/runner-images/issues/6002
* `macos-10.15` : https://github.com/actions/runner-images/issues/5583

Due to this there was failing tests due to planned brownouts.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add license detection rules

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Bump versions to 32.0.0b1

Also update spdx license list version as it was updated previously.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix test failures

- Update test fails due to spdx license list update
- Update test fails due to inconsistent rule data

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra commented:
@pombredanne I've fixed the test failures, but let's keep this open as discussed for the two changes i.e. top level unique license detections and license references as default, I'll open PRs for those into this branch/merge those onto develop and then merge this just before pushing the `32.0.0b1` tag?

AyanSinhaMahapatra committed:
Fix license/rules dump format

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'develop' into release-32-rc1-prep

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Reformat licenses with a load/dump

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Reformat rules with a load/dump

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add license updates from spdx license list 3.19

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra commented:
@DennisClark I would need your review on the following commit: https://github.com/nexB/scancode-toolkit/pull/3150/commits/088e10668c677734798d36a8c0ff862f17939491 i.e. license additions in scancode-toolkit from spdx license list 3.19

AyanSinhaMahapatra commented on src/licensedcode/data/licenses/x11vnc-openssl-exception.LICENSE:
```diff
@@ -0,0 +1,21 @@
+    - https://github.com/LibVNC/x11vnc/blob/master/src/8to24.c#L22
+---
+
+In addition, as a special exception, Karl J. Runge
```
Is this just https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/licenses/openssl-exception-gpl-2.0.LICENSE ?

scan result:

```
 {
      "path": "sca60",
      "type": "file",
      "licenses": [
        {
          "key": "generic-exception",
          "score": 99.0,
          "name": "Generic Exception Notice",
          "short_name": "Generic Exception Notice",
          "category": "Unstated License",
          "is_exception": true,
          "is_unknown": false,
          "owner": "Unspecified",
          "homepage_url": null,
          "text_url": "",
          "reference_url": "https://scancode-licensedb.aboutcode.org/generic-exception",
          "scancode_text_url": "https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/licenses/generic-exception.LICENSE",
          "scancode_data_url": "https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/licenses/generic-exception.yml",
          "spdx_license_key": "LicenseRef-scancode-generic-exception",
          "spdx_url": "https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/licenses/generic-exception.LICENSE",
          "start_line": 2,
          "end_line": 2,
          "matched_rule": {
            "identifier": "license-intro_8.RULE",
            "license_expression": "generic-exception",
            "licenses": [
              "generic-exception"
            ],
            "referenced_filenames": [],
            "is_license_text": false,
            "is_license_notice": false,
            "is_license_reference": false,
            "is_license_tag": false,
            "is_license_intro": true,
            "has_unknown": false,
            "matcher": "2-aho",
            "rule_length": 3,
            "matched_length": 3,
            "match_coverage": 100.0,
            "rule_relevance": 99
          },
          "matched_text": "as a special exception,"
        },
        {
          "key": "openssl-exception-gpl-2.0",
          "score": 90.48,
          "name": "OpenSSL exception to GPL 2.0",
          "short_name": "OpenSSL exception to GPL 2.0",
          "category": "Copyleft",
          "is_exception": true,
          "is_unknown": false,
          "owner": "OpenSSL",
          "homepage_url": "http://www.openssl.org/source/license.html",
          "text_url": "http://www.openssl.org/source/license.html",
          "reference_url": "https://scancode-licensedb.aboutcode.org/openssl-exception-gpl-2.0",
          "scancode_text_url": "https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/licenses/openssl-exception-gpl-2.0.LICENSE",
          "scancode_data_url": "https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/licenses/openssl-exception-gpl-2.0.yml",
          "spdx_license_key": "LicenseRef-scancode-openssl-exception-gpl-2.0",
          "spdx_url": "https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/licenses/openssl-exception-gpl-2.0.LICENSE",
          "start_line": 3,
          "end_line": 10,
          "matched_rule": {
            "identifier": "openssl-exception-gpl-2.0_1.RULE",
            "license_expression": "openssl-exception-gpl-2.0",
            "licenses": [
              "openssl-exception-gpl-2.0"
            ],
            "referenced_filenames": [],
            "is_license_text": true,
            "is_license_notice": false,
            "is_license_reference": false,
            "is_license_tag": false,
            "is_license_intro": false,
            "has_unknown": false,
            "matcher": "3-seq",
            "rule_length": 105,
            "matched_length": 95,
            "match_coverage": 90.48,
            "rule_relevance": 100
          },
          "matched_text": "gives permission to link the code of its release [of] [x11vnc] with the\nOpenSSL project's \"OpenSSL\" library (or with modified versions of it\nthat use the same license as the \"OpenSSL\" library), and distribute\nthe linked executables.  You must obey the GNU General Public License\nin all respects for all of the code used other than \"OpenSSL\".  If you\nmodify this file, you may extend this exception to your version of the\nfile, but you are not obligated to do so.  If you do not wish to do\nso, delete this exception statement from your version."
        }
      ],
      "license_expressions": [
        "generic-exception",
        "openssl-exception-gpl-2.0"
      ],
      "percentage_of_license_text": 93.33,
      "scan_errors": []
    }
  ]
}
```

DennisClark commented:
@AyanSinhaMahapatra   you are right, thanks.  the spdx x11vnc license is the very same as our openssl-exception-gpl-2.0 if you follow the SPDX matching guidelines.  So we need to update the SPDX identifier of our license with x11vnc-openssl-exception  instead of the current licenseref value and not create a new license.

AyanSinhaMahapatra committed:
Vendor python-forntmatter completely

We used to vendor parts of python frontmatter and also use it as
a dependency which was problematic. This vendors python-frontmatter
fully and only uses the code needed.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update synclic script and misc test expectations

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add new license category "CLA" #3147 #3038

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Deprecate licenses and rules

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add new and updated licenses and rules

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix test failures

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'license-detection-follow-up' into release-32-rc1-prep

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne committed:
Do not use packaging.LegacyVersion #3171 #3177

* Switch to using packvers fork of packaging
* Bump pip requirements parser, but it still using packaging.
* Bump dparse2 which is using packvers
* Pin SPDX tools for cope with breaking APIs
* Update release script to build one release wheel for each Python
  version that each contain a pickle of the License index
* Streamline the app build
* Bump version and use correct SPDX version

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

AyanSinhaMahapatra committed:
Update debian and spdx output plugins

* fix bug in debian output plugin
* update plugins to not fetch matched_text from reference data

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add license dump CLI option back

* the license dump CLI option was removed by mistake in a merge
  conflict resolution, this adds it back.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add new docstrings and update old docstrings

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update rule_identifiers for synthetic rules

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update CHANGELOG

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update test expecations for recent changes

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update documentation

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Ensure license dump test does not regen footer

* Align to other test expectations

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Remove dangling code using mappings

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix test expectations

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'release-32-rc1-prep' into v31.2.2-branch-hotfix

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge pull request #3180 from nexB/v31.2.2-branch-hotfix

Do not use packaging.LegacyVersion #3171 #3177

pombredanne commented on src/licensedcode/models.py:
```diff
@@ -1950,6 +2008,18 @@ def _from_expression(cls, license_expression=None, identifier=None, **kwargs):
+            text = "None"
+        else:
+            text = self.text
+        return hashlib.md5(text.encode('utf-8')).hexdigest()
```
Please enter an issue to fix the md5 problem for #3165 FIPS support

pombredanne replied:
I updated this in the latest push to use sha1. @RayGozer

pombredanne commented on src/licensedcode/models.py:
```diff
@@ -1950,6 +2008,18 @@ def _from_expression(cls, license_expression=None, identifier=None, **kwargs):
+        an MD5 checksum of the text, but that's an implementation detail)
+        """
+        if not self.text:
+            text = "None"
```
We cannot have all the items without a text having the same id. You need to inject something else then IMHO, unless you are sure this will not be an issue. But then a UUID may be better?

pombredanne replied:
I updated the code to be:
```
        content = f'{self.identifier!r}{self.license_expression!r}{self.text!r}'
        return sha1(content.encode('utf-8')).hexdigest()
```

pombredanne commented on src/packagedcode/alpine.py:
```diff
@@ -1647,7 +1647,7 @@ def apply_syntax_fixes(s):
     # "with" mishaps
     'gpl-2.0 with classpath': 'gpl-2.0 with classpath-exception-2.0',
-    'gpl-2.0-only with openssl-exception': 'gpl-2.0-only with licenseref-scancode-openssl-exception-gpl-2.0',
+    'gpl-2.0-only with openssl-exception': 'x11vnc-openssl-exception',
```
Are you sure? I do not think this is right.

pombredanne replied:
I fixed this

pombredanne commented on src/scancode/api.py:
```diff
@@ -144,7 +144,10 @@ def get_urls(location, threshold=50, **kwargs):
 DEJACODE_LICENSE_URL = 'https://enterprise.dejacode.com/urn/urn:dje:license:{}'
 SCANCODE_LICENSEDB_URL = 'https://scancode-licensedb.aboutcode.org/{}'
-
+SCANCODE_DATA_BASE_URL = 'https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data'
```
We will need to include RULEs in the license DB soon enough, FYI.

pombredanne replied:
See https://github.com/nexB/scancode-toolkit/issues/3183

pombredanne reviewed (changes requested):
Thanks!
Here are some review comments

pombredanne committed:
Add new BSD simplified variant

And use correct expectation for license test

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Format license text and remove copyright notice

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Format license text and remove copyright notice

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Restore correct test expectation

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Align ignorables with copyright removal from text

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Generate correct "rule file" paths

These are used mostly for display in test failures.
Synthetic rules such as SPDX have no backing files and LICENSE and RULE
files use their file name as Rule.identifier. The new SynthethicRule
Rule subclass is used now as a base for these rules.
Also format code.


Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Format code

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne reviewed (approved):
LGTM! Thank you ++

pombredanne committed:
Merge remote-tracking branch 'upstream/develop' into release-32-rc1-prep

pombredanne committed:
Streamline license_db dump command

Improve docstrings, add ount to CLI output and redirect to stderr.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Bump to 32.0.0rc1

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Remove duplicate class

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Organize imports

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Guard tracing with TRACE flag

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Apply cosmetic refactorings

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Remove duplicated function

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Remove duplicated import

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Guard tracing with TRACE flag

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Streamline imports

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Streamline imports

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Streamline imports

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Streamline imports

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Streamline imports

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Streamline imports and unused variables

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Use correct Package factory

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Remove unused code

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Use correct function name

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Rename function name to avoid duplication

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Streamline imports

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Streamline imports

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Format code

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Streamline imports

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Streamline imports

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Guard imports from cache inside functions

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Add support for FIPS

Reported-by: RayGozer @RayGozer
Reference: https://github.com/nexB/scancode-toolkit/issues/3165
Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Ensue rule id is unique also for rule without texts

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Restore correct license expression mapping

gpl-2.0-only with licenseref-scancode-openssl-exception-gpl-2.0
is NOT about x11vnc

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Import sys at the top

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Add support for FIPS

Use sha1 correctly for UUID seed

Reported-by: RayGozer @RayGozer
Reference: https://github.com/nexB/scancode-toolkit/issues/3165
Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Regen test with new rule identifier

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Update tests expectations

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Apply cosmetic refactoring

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve Rule check with is_synthetic check

Also move test utilities to test module

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Use correct imports in tests

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Avoit using copyright as a varibale name

Prefer copyrght to avoid name collision with the built-in
copyright

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Rename and streamline get_rule_object_from_match

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Make license reference its own plugin

It kicks in if license or package are requested as a post-scan plugin

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Remove deprecated is_license_text plugin

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Use correct planned release date

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve license and rule references handling

Streamline code now that we do not deal with references in the main
detection.

Improve models accordingly

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Do not include SPDX expression in low level matches

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Update test expectations for RPMs

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Update test expectations

Align with latest fixesa nd the move of license details to
license references

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Do not keep is_deprecated License flag

No deprecated license can participate in detection.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Enable LicenseReference plugin correctly

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Update test expectations

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Merge upstream branch update-ci-images

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Generate correct Rule URL

Additional rules and synthetic rules have no URL for now.
The logic to create such URL is now under Rule.rule_url

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Use plain JSON test file

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne commented:
@AyanSinhaMahapatra this is ready for your review. The key changes are:
- the license references are now always created with a new separate post scan plugin. This plugin is not commanded by its own command line option. Instead it is enabled if either the license or package option is requested. Because of this I streamlined or removed quite a bit of license reference-related code. 
- I updated which fields are included in the license and rule references to include many more fields.
- the package license detection do not contain any license/rule references anymore, this is now only top level.
- I removed the deprecated "is_license_text" plugin
- I replaced using md5 with sha1 for #3165 and eventual future FIPS support
- I applied cosmetic refactoring and imports cleanup across the codebase

pombredanne committed:
Correct license mappings for Alpine

We were missing out certain openssl exception

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Generate correct URLs for license and rules

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Use correct URLs for license and rules

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne commented:
Back to green! I am merging this now!
